### PR TITLE
Add support for Waveshare 7.5inch color e-paper display

### LINKED ISSUE #239: Support for  Waveshare Three-color E-Ink Displays
**Describe the problem you have/What new integration you would like**

References #77. 
I've got Waveshare's 3-color e-ink display and it would be nice to have some support for them.

**Please describe your use case for this integration and alternatives you've tried:**

**Additional context**

### PR DESCRIPTION
**Pull request by asmfreak**
## Description:

**Related issue (if applicable):** partially fixes esphome/feature-requests#239

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#286

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/waveshare_epaper/flash_array.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+};
+
+template<unsigned int id, size_t N, size_t buffer_size_bytes>
+void
+BufferedFlashArray<id, N, buffer_size_bytes>::commit(){
+  while(dirty){
+    // wavesh_log(" dirty %x from %x", (aligned_type)(arr), (aligned_type)(small_buffer_start) );
+    noInterrupts();
+    auto res = spi_flash_erase_sector(
+      reinterpret_cast<aligned_type>(small_buffer_start)/SPI_FLASH_SEC_SIZE);
```
Comment (OttoWinter) [inline esphome/components/waveshare_epaper/flash_array.h]: No no no, this is not ok.

1. Writing the whole buffer to flash will absolutely kill the flash in a short amount of time.

2. This API is specific to ESP8266 and will not work for ESP32, even though on ESP32 this integration would be much better. If anything, we should make this an ESP32-only integration.

3. As you say in the docs, writing to the display this way takes ages - that will slow down ESPHome's core loop enough to cause lots of problems. Including for example dropping native API connections.

This approach of using flash for the display storage is a big red flag.

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp [hunk 2]
```diff
@@ -596,6 +595,170 @@ void WaveshareEPaper7P5In::dump_config() {
   LOG_PIN("  Busy Pin: ", this->busy_pin_);
   LOG_UPDATE_INTERVAL(this);
 }
+const uint8_t COLOR_BLACK{display::COLOR_ON};
+const uint8_t COLOR_YELLOW{(uint8_t)(COLOR_BLACK + 1)};
+const uint8_t COLOR_RED{COLOR_YELLOW};
+const uint8_t COLOR_WHITE{display::COLOR_OFF};
```
Comment (OttoWinter) [inline esphome/components/waveshare_epaper/waveshare_epaper.cpp]: Please wait for color support in display base abstraction and don't fake it here. See the esphome-core PR for display colors. First, the base needs to be updated for color support, only then are integrations allowed to use that.

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp [hunk 3]
```diff
@@ -596,6 +595,170 @@ void WaveshareEPaper7P5In::dump_config() {
...
+
+void wavesh_log(const char* format, ...){
+  va_list arg;
+  va_start(arg, format);
+  const static char *TAGG = "wshe" ;
+  char buffer[200];
+  vsnprintf(buffer, 200, format, arg);
+  va_end(arg);
+  ESP_LOGI(TAGG, "%s ", buffer);
```
Comment (OttoWinter) [inline esphome/components/waveshare_epaper/waveshare_epaper.cpp]: Why this function? You know you can just use a different TAG variable if you want to have a different prefix for some log messages.

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp [hunk 4]
```diff
@@ -596,6 +595,170 @@ void WaveshareEPaper7P5In::dump_config() {
...
+  const uint8_t fill = (color == COLOR_WHITE) ? 0x00 :
+    ((color == COLOR_YELLOW)?0b10101010: 0b01010101);
+  ESP_LOGD(TAG, "Fill start");
+  for (uint32_t i = 0; i < this->static_buffer_length_(); i++){
+    this->static_buffer_.write(i,fill);
+    App.feed_wdt();
+  }
+  ESP_LOGD(TAG, "Fill end");
+}
+uint32_t WaveshareEPaper7P5InC::get_buffer_length_(){ return 1; }
```
Comment (OttoWinter) [inline esphome/components/waveshare_epaper/waveshare_epaper.cpp]: This is a red flag too - don't abuse the internal APIs just to hack things to get it working.

**Review by OttoWinter — COMMENTED** [review]
I'm sorry but this PR is not in any shape to be merged.

The whole code just feels way to hacky to be accepted as official code (for example `BufferedFlashArray`), and there's commented out lines and warnings as debug logs littered everywhere.

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp (hunk 3 above)
Comment (ASMfreaK) [reply esphome/components/waveshare_epaper/waveshare_epaper.cpp]: This was a debug function for template classб Sorry for messed-up debug prints. I just wanted to start process on this.

File: esphome/components/waveshare_epaper/flash_array.h (hunk 1 above)
Comment (ASMfreaK) [reply esphome/components/waveshare_epaper/flash_array.h]: 1. FLASH buffer was just a proof-of-concept. I want to move to SPI RAM chip in the future.
2. The same goes here. I'm against making it ESP32-only, as I own ESP8266 module for display.
3. I'd do the writing asyncronously, if I knew some way. ESPHome doesn't have some kind of `processEvents()` function like Qt? So, if I have a long operation,other tasks would have a chance to execute?

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp (hunk 4 above)
Comment (ASMfreaK) [reply esphome/components/waveshare_epaper/waveshare_epaper.cpp]: This is not an abuse, but a workaround. There is no way (without major reworking) to use DisplayBuffer WITHOUT built-in buffer. I'm quite sure, there are displays which can have bigger sizes then this one.
Do tell me how can I use an external buffer for a display WITHOUT disabling this built-in buffer?

**Pull request closed without merging**